### PR TITLE
Add a flag for background compilation

### PR DESCRIPTION
**Pull request by abarth**
This flag won't do anything until the next engine roll.

### CODE REVIEW

**Comment by abarth** [comment]
@devoncarew @eseidelGoogle

File: packages/flutter_tools/lib/src/android/device_android.dart [hunk 1]
```diff
@@ -283,12 +284,15 @@ class AndroidDevice extends Device {
...
       cmd.addAll(<String>['--ez', 'enable-checked-mode', 'true']);
+    if (backgroundCompilation)
+      cmd.addAll(<String>['--ez', 'enable-background-compilation', 'true']);
     if (traceStartup)
       cmd.addAll(<String>['--ez', 'trace-startup', 'true']);
     if (startPaused)
       cmd.addAll(<String>['--ez', 'start-paused', 'true']);
     if (route != null)
       cmd.addAll(<String>['--es', 'route', route]);
+    print('cmd=$cmd');
```
Comment (eseidelGoogle) [inline packages/flutter_tools/lib/src/android/device_android.dart]: Did you mean to leave this?

**Comment by eseidelGoogle** [comment]
lgtm, see nit above.

File: packages/flutter_tools/lib/src/android/device_android.dart (hunk 1 above)
Comment (abarth) [inline packages/flutter_tools/lib/src/android/device_android.dart]: Nope!

File: packages/flutter_tools/lib/src/commands/start.dart [hunk 2]
```diff
@@ -38,6 +38,9 @@ abstract class StartCommandBase extends FlutterCommand {
         negatable: true,
         defaultsTo: true,
         help: 'Toggle Dart\'s checked mode.');
+    argParser.addFlag('enable-background-compilation',
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/commands/start.dart]: Is this something we want to show to users? I'm guessing this is something we'll experiment w/ for a bit, then enable by default in the engine.

**Comment by devoncarew** [comment]
lgtm. It's eye-opening to see how many places there are to modify to pass a single flag through.

You'll want to add this to the ios subclasses as well (as TODO:s).

File: packages/flutter_tools/lib/src/commands/start.dart (hunk 2 above)
Comment (abarth) [inline packages/flutter_tools/lib/src/commands/start.dart]: Correct.  I'll hide it.

**Comment by abarth** [comment]
> It's eye-opening to see how many places there are to modify to pass a single flag through.

Yeah, we might want to extract a Config object of some sort.

> You'll want to add this to the ios subclasses as well (as TODO:s).

It doesn't make sense on iOS because we use AOT compilation there.

**Comment by devoncarew** [comment]
> It doesn't make sense on iOS because we use AOT compilation there.

Ah, ok, makes sense. There are two iOS subclasses of Device however. You'll want to add no-op named arg to the implementations.

You could also pass the param in the `platformArgs` map. That's where we put one or two android specific params that don't makes sense as a general named param into Device. It sounds about right for `backgroundCompilation`.

**Comment by abarth** [comment]
Not flag needed.  The tool does this by default now.

**Pull request closed without merging**